tests/unit/test_metrics.py

```diff
@@ -216,3 +222,165 @@ def test_mmd_random_data(
+
+        # Generate data from a uniform distribution Q with which to compute the expected
+        # KSD.
+        comparison_pts = jr.uniform(
+            random_key,
+            shape=(num_data_points, dim),
+        )
+
+        # Analytic stein kernel
+        def _stein_kernel(x: ArrayLike, y: ArrayLike) -> Array:
+            norm_sq = jnp.linalg.norm(jnp.subtract(x, y)) ** 2
+            body = 1 + norm_sq
+            first_term = 3 * norm_sq / body ** (5 / 2)
+            second_term = (dim - norm_sq) / body ** (3 / 2)
+            third_term = jnp.dot(x, y) / body ** (1 / 2)
+            return -first_term + second_term + third_term
+
+        stein_kernel_matrix = pairwise(_stein_kernel)(comparison_pts, comparison_pts)
+        expected_output = stein_kernel_matrix.sum() / num_data_points**2
+
+        # Compute the KSD using the metric object via a stein kernel induced by a
+        # PCIMQ kernel, with the corresponding analytic score function and
+        # reference data from a standard multivariate distribution P.
+        reference_pts = jr.multivariate_normal(
+            random_key,
+            jnp.zeros(dim),
+            jnp.eye(dim),
+            shape=(num_data_points,),
+        )
+        base_kernel = PCIMQKernel(length_scale=1 / np.sqrt(2), output_scale=1)
+
+        def _score_function(x: ArrayLike) -> Array:
+            return -jnp.asarray(x)
+
+        stein_kernel = SteinKernel(base_kernel, score_function=_score_function)
+        metric = KSD(kernel=stein_kernel)
+        output = metric.compute(Data(reference_pts), Data(comparison_pts))
+
+        assert output == pytest.approx(expected_output)
+
+    @pytest.mark.parametrize(
+        "mode", ["unweighted", "weighted", "laplace-corrected", "regularised"]
+    )
+    def test_ksd_random_data(
+        self,
+        problem: _MetricProblem,
+        mode: Literal["unweighted", "weighted", "laplace-corrected", "regularised"],
+    ):
+        r"""
+        Test KSD computed from randomly generated test data agrees with method result.
+
+        - "unweighted" parameterization checks that if the 'reference_data' and the
+            'comparison_data' have the default 'None' weights, that the computed KSD
+              is
+            given by the means of the unweighted kernel matrices.
+        - "weighted" parameterization checks that for arbitrarily weighted data, the
+            computed MMD is given by the weighted average of the kernel matrices.
+        """
+        x, y = problem
+
+        base_kernel = SquaredExponentialKernel()
+        kernel = convert_stein_kernel(x.data, base_kernel, None)
+        metric = KSD(kernel=kernel, score_matching=None)
+
+        # Compute each term in the KSD formula to obtain an expected KSD.
+        kernel_mm = kernel.compute(y.data, y.data)
+        if mode == "weighted":
+            weights_mm = y.weights[..., None] * y.weights[None, ...]
+            expected_ksd = jnp.average(kernel_mm, weights=weights_mm)
```

Comment:
Is there a reason we use `jnp.average` here but `jnp.mean` below?

Reply:
I use `jnp.average` in order to make correct use of the weights, below we do an unweighted mean so no need for it.